sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_session_async.py

```diff
@@ -414,7 +414,8 @@ async def end(self, error=None, wait=False):
             if self.state not in [SessionState.UNMAPPED, SessionState.DISCARDING]:
                 await self._outgoing_end(error=error)
             for _, link in self.links.items():
-                await link.detach()
+                if link._is_closed is False: # pylint: disable=protected-access
```

Comment:
Looking at the link detach code, it looks like `close=False` by default, which means calling `link.detach()` without parameters wont result in `is_closed` being set to True.
So it seems that the original intention behing `is_closed` was _not_ to be used interchangably with `is_detached`, which is how it's used here.

Reply:
Whether this requires a more extensive audit of the use of `_is_closed` or not, I'm not sure.
It is an internal attribute to `link` - so should not be consumed by Session in this manner. Though I'm not sure if anything else is also consuming it like this.
If we do need to expose some kind of access to the Session object, we should add a public attribute/method for it.

Reply:
yeah that was the conundrum as `_is_closed` is used even at the client [level](https://github.com/Azure/azure-sdk-for-python/blob/6f4abdd4a870a9ef33981d9c3ca7d18ceffdbc75/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/client.py#L234). 

I was wondering instead of adding an attribute on `Link`, adding something like `detach_link(link_name)` method on Session. The links created by client and mgmt are done via `Session` methods on Session (`create_receiver_link`, `create_sender_link`, `create_request_response_link_pair`) , so detach should also be done at that level ?

Reply:
thankfully its usage is outside of `link` is limited to `session`, `client` and `receiver` and that too sparingly.  Given that service bus doesn't support link recovery, I'm guessing the use was to determine if the link was in a detaching state and not whether its closed. Ive added a `is_detaching` method and maybe `is_closed` or ( stopped ) is something to expose for the future.